src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,134 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] dataArray;
+    private int size;
+
+    public ArrayList(int capacity) {
+        if (capacity >= 0) {
+            this.dataArray = new Object[capacity];
+        } else {
+            throw new IllegalArgumentException("Capacity can't be negative number");
+        }
+    }
+
+    public ArrayList() {
+        this.dataArray = new Object[DEFAULT_CAPACITY];
+    }
 
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
+        dataArray[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        indexCheckForAdd(index);
+        ensureCapacity(size + 1);
+        System.arraycopy(dataArray, index, dataArray, index + 1, size - index);
+        dataArray[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        Object[] listArray = list.toArray();
+        ensureCapacity(size + listArray.length);
+        System.arraycopy(listArray, 0, dataArray, size, listArray.length);
+        size += listArray.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return (T) dataArray[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexCheck(index);
+        dataArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        T removedValue = (T) dataArray[index];
+        fastRemove(index);
+        return removedValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        if (element == null) {
+            for (int i = 0; i < size; i++) {
+                if (dataArray[i] == null) {
+                    fastRemove(i);
+                    return null;
+                }
+            }
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (element.equals(dataArray[i])) {
+                    T removedValue = (T) dataArray[i];
+                    fastRemove(i);
+                    return removedValue;
+                }
+            }
+        }
```

Comment:
why should we use if-else blocks here? maybe better simplify it?

Reply:
done